### ISSUE TITLE
`cargo fmt` removes enum variants

### ISSUE THREAD

**Issue by jamesmunns** [rust-lang-nursery/rustfmt#2479]
Hello, I am using `cargo fmt` installed as described in the readme.txt. When attempting to format code generated by @japaric 's `svd2rust` tool, I get the following snippet (it is one really painfully long line):

```rust

# [ derive ( Clone , Copy , Debug , PartialEq ) ] pub enum POLARITYR {    # [ doc = "Task mode: No effect on pin from OUT[n] task. Event mode: no IN[n] event generated on pin activity." ] NONE , # [ doc = "Task mode: Set pin from OUT[n] task. Event mode: Generate IN[n] event when rising edge on pin." ] LOTOHI , # [ doc = "Task mode: Clear pin from OUT[n] task. Event mode: Generate IN[n] event when falling edge on pin." ] HITOLO , # [ doc = "Task mode: Toggle pin from OUT[n]. Event mode: Generate IN[n] when any change on pin." ] TOGGLE }

```

When running this through `cargo fmt`, I get the following output:

```rust
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum POLARITYR {}
```

While the correct output should be something like this:

```rust
# [ derive ( Clone , Copy , Debug , PartialEq ) ]
pub enum POLARITYR {
    # [doc = "Task mode: No effect on pin from OUT[n] task. Event mode: no IN[n] event generated on pin activity." ]
    NONE,

    # [ doc = "Task mode: Set pin from OUT[n] task. Event mode: Generate IN[n] event when rising edge on pin." ]
    LOTOHI, 

    # [ doc = "Task mode: Clear pin from OUT[n] task. Event mode: Generate IN[n] event when falling edge on pin." ]
    HITOLO,

    # [ doc = "Task mode: Toggle pin from OUT[n]. Event mode: Generate IN[n] when any change on pin." ]
    TOGGLE
}
```

Because the Enum Variants are omitted, this causes code problems when compiling. I am not sure if other parts of the formatting are affected.

**Comment by jamesmunns** [comment]
I actually think this might have to do with the length of the doc string. Consider the following results:

Input:
```rust
# [ derive ( Clone , Copy , Debug , PartialEq ) ] pub enum POLARITYR {     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     NONE,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     LOTOHI,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     HITOLO,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     TOGGLE }
```

Output (correct):
```rust
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum POLARITYR {
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa "] NONE,
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa "] LOTOHI,
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa "] HITOLO,
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa "] TOGGLE,
}
```

Input:
```rust
# [ derive ( Clone , Copy , Debug , PartialEq ) ] pub enum POLARITYR {     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     NONE,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     LOTOHI,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     HITOLO,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa " ]     TOGGLE }
```

Output (failed):
```rust
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum POLARITYR {}
```

**Comment by jamesmunns** [comment]
Even more interesting, at 87 characters, you get a warning, but no failure.

```rust
# [ derive ( Clone , Copy , Debug , PartialEq ) ] pub enum POLARITYR {     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa" ]     NONE,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa" ]     LOTOHI,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa" ]     HITOLO,     # [ doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa" ]     TOGGLE }
```

```text
error: line exceeded maximum width (maximum: 100, found: 101)
 --> /home/james/personal/nrf52/src/lib.rs:3
  |
3 |     #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
  |                                                                                                     ^

error: line exceeded maximum width (maximum: 100, found: 101)
 --> /home/james/personal/nrf52/src/lib.rs:5
  |
5 |     #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
  |                                                                                                     ^

error: line exceeded maximum width (maximum: 100, found: 101)
 --> /home/james/personal/nrf52/src/lib.rs:7
  |
7 |     #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
  |                                                                                                     ^

error: line exceeded maximum width (maximum: 100, found: 101)
 --> /home/james/personal/nrf52/src/lib.rs:9
  |
9 |     #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
  |                                                                                                     ^

warning: rustfmt may have failed to format. See previous 4 errors.
```

```rust
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum POLARITYR {
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
    NONE,
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
    LOTOHI,
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
    HITOLO,
    #[doc = "aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaaa aaaaaaaaa"]
    TOGGLE,
}
```

So,

* At <=86 characters, everything is fine
* At == 87 characters, you get an "error", but the output is still correct.
* At >= 88 characters, you get NO error, and the output is incorrect.

**Comment by topecongiro** [comment]
@jamesmunns Thank you for a detailed issue report!

One thing we can do in rustfmt is to convert `#[doc = "doc comment"]` to `/// doc comment`. AFAIK these two are semantically equivalent, the only difference being `#[doc = "..."]` makes it easier for macros to produce doc comments.

**Comment by topecongiro** [comment]
This is somewhat fixed as a side effect of #2471. Enum variants will not get removed anymore, though some attributes are formatted poorly.

```rust
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum POLARITYR {
    # [ doc = "Task mode: No effect on pin from OUT[n] task. Event mode: no IN[n] event generated on pin activity." ]
    NONE,
    # [ doc = "Task mode: Set pin from OUT[n] task. Event mode: Generate IN[n] event when rising edge on pin." ]
    LOTOHI,
    # [ doc = "Task mode: Clear pin from OUT[n] task. Event mode: Generate IN[n] event when falling edge on pin." ]
    HITOLO,
    #[doc = "Task mode: Toggle pin from OUT[n]. Event mode: Generate IN[n] when any change on pin."]
    TOGGLE,
}
```
